### ISSUE TITLE
Support SpringAware on MapStore

### ISSUE THREAD

**Issue by nkavian** [hazelcast/hazelcast#18555]
**Please describe the problem you are trying to solve**
Users of Spring are now accustomed that annotation abased configuration is the norm. It seems Hazelcast-Spring forces configuration via XML file. In addition, Hazelcast YML is not supported with Hazelcast Spring.

In order to use MapStore with an autowired member, you must use the XML Spring context XML file mechanism. This has the consequence a developer can not use Java + annotation based configuration. Notably, I need to be able to set certain Config variables and properties after Spring has loaded it's application.properties and after I have access to server environment variables.

Neither configuration mechanism can solely be used in my project.

**Please describe the desired behavior**
Hazelcast Spring should detect and honor Spring dependency injection on MapStore classes annotated with `@SpringAware`.

**Describe alternatives you've considered**
I've spent 3 days going through bug reports, Google, and Github looking for a solution. The Hazelcast Spring documentation is outdated and incomplete. I can only offer this repo I've updated to reflect the issue as unit tests.

https://github.com/nkavian/springdemo2
mvn clean test

There is an XML project that shows MapStore working.
There is a Java Annotation project that shows MapStore+SpringAware not working.

**Comment by Holmistr** [comment]
Hi @nkavian , thanks a lot for opening this issue.

@ldziedziul @frant-hartm Could I ask you to take a quick look and asses what is the complexity of this?

**Comment by ldziedziul** [comment]
I think we need to check first how we handle the map configuration as a part of the Spring JavaConfig and annotations. From a very quick glimpse of an eye, it seems we support only XML configuration. So to be honest, the issue could grow from supporting MapStore configuration to better support of the configuration hazelcast itself using spring, not sure right now.

My current and naive understanding would be that it's not a trivial change and require some research first to understand the scope of the work. (although I think it's required as JavaConfig and annotation are far more popular nowadays than XML configs)

**Comment by frant-hartm** [comment]
@nkavian Thanks for the report and the reproducer.

Your XML and Java configurations are not identical. Your Java config is missing the MapStore configuration.

```
   @Bean
   public Config createConfigWithSpringManagedContext(ApplicationContext context, SpringStore springStore) {
       Config config =  new Config();

       MapStoreConfig mapStoreConfig = new MapStoreConfig();
       mapStoreConfig.setImplementation(springStore);
       mapStoreConfig.setWriteDelaySeconds(15);
       config.getMapConfig("aMap").setMapStoreConfig(mapStoreConfig);

       SpringManagedContext springManagedContext = new SpringManagedContext();
       springManagedContext.setApplicationContext(context);
       config.setManagedContext(springManagedContext);
       return config;
   }
```

It makes the test in your project pass. 

However, I agree there are multiple issues 
- the documentation for this particular feature should be clearer
- in general, we don't have docs on how Hazelcast should be configured in Spring (and Spring Boot) with Java config, and there are probably some gaps
- using Hazelcast XML config/Yaml config (or when you use `setClassName` instead of `setImplementation`) for the MapStore with SpringAware doesn't work

**Comment by nkavian** [comment]
@frant-hartm agreed, they are not identical. I created the ticket and sample repo before I found+implemented workarounds; and later gained a deeper understanding of how to configure it without documentation.

In my production code, I use `hazelcast.yml` to configure the basic settings of the map store, and then I attached a `@SpringAware` instance like this: `config.getMapConfig("aMap").getMapStoreConfig().setImplementation(new AMapStore());`

Your 3 bullet points are a very clear recap of the issues.

I like your example of a `@Component` SpringStore, I'll give that another try and see if that resolves a cyclic dependency and Spring hang.
